test/functorch/test_control_flow.py

```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+        self.assertEqual(result_out, expected_result_out)
+
+        # With init
+        inp = (x, h)
+        init = (torch.zeros_like(x[:, 0:1, :]), torch.clone(h[:, 0:1, :]))
+        result = scan(RNN, inp, dim=dim, init=init)
+        result_out = result[1][:, 1:, :]
+        expected_result = rnn(
+            torch.permute(x, (1, 0, 2)), torch.unsqueeze(h[:, 0, :], 0)
+        )
+        expected_result_out = torch.permute(expected_result[0], (1, 0, 2))
+        self.assertEqual(result_out, expected_result_out)
+
+    @unittest.skipIf(not SM70OrLater, "triton")
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @parametrize("reverse", [False, True])
+    @parametrize("compile_mode", ["none", "eager", "compile", "compile_dynamic_shape"])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    # Skipping the scan tests, as scan currently does not support inductor
+    @decorateIf(
+        unittest.skip,
+        lambda params: (params["compile_mode"] != "eager"),
+    )
+    def test_scan_init(self, reverse, compile_mode, device):
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        torch.compiler.reset()
+        if compile_mode == "compile":
+            scan_fct = torch.compile(scan, fullgraph=True, dynamic=False)
+        elif compile_mode == "compile_dynamic_shape":
+            scan_fct = torch.compile(scan, fullgraph=True, dynamic=True)
+        elif compile_mode == "eager":
+            scan_fct = torch.compile(scan, fullgraph=True, backend="eager")
+        else:
+            scan_fct = scan
+
+        # Only init and no input
+        x = torch.randn(3, 1, 2, device=device)
+        init = torch.randn(3, 1, 2, device=device)
+        dim = 1
+        op, op_pt = (add, torch.cumsum)
+
+        result = scan_fct(op, input=x, dim=dim, reverse=reverse)
+        result_exp = _fake_scan(op, x, dim=dim, reverse=reverse)
+
+        # Scan dimension is 0
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Only init given
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        result_init = scan_fct(op, input=[], dim=dim, reverse=reverse, init=init)
+        self.assertEqual(result, result_exp)
+        self.assertEqual(result_init, result_exp)
+        self.assertEqual(result_init, init)
+
+        x = torch.randn(3, 1, 2, device=device)
+        init = torch.randn(3, 1, 2, device=device)
+        dim = 1
+
+        op, op_pt = (add, torch.cumsum)
+        result = scan_fct(op, input=x, dim=dim, reverse=reverse)
+
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        # Init scalar
+        init = 1.0
+        with self.assertRaisesRegex(Exception, ".*"):
```

Comment:
Testing exceptions should be in another test. Also please don't just catch all Exception (should catch specific action with proper error message). Otherwise, it could cause silent errors.

Reply:
Right, I fully agree that catching generic Exceptions is bad, but I couldn't really figure out a way how to do that properly. The problem is that the AssertionError are masked by some errors coming from dynamo, e.g., 

```
torch._dynamo.exc.Unsupported: should_compile_partial_graph=False

from user code:
   File "/data_malta3_ssd/pytorch/torch/_higher_order_ops/scan.py", line 158, in scan
    assert all(
```

Do you have any suggestions how to properly do this?

Reply:
Can you show me the full error message? I guess that's because of a graph break. This is because dynamo will inline the code in scan.py and somewhere in `assert all(...)` is causing a graph break. Because we torch.compile(fullgraph=True) hence get this error. 


We could either fix the graph break in dynamo (if we can) or try to use an supported approach to do the same thing (e.g. can you try raise an runtime error or something?)

Reply:
I see. For the `cond.py` we had changed the `asserts` into `RuntimeError` which where supported and hence could be properly catched in the testcase. I did the same here and also aligned the respective cases in `associative_scan`. As discussed, if dynamo breaks the graph due to error, the `assertRaisesRegex` will catch `torch._dynamo.exc.Unsupported, "Observed exception.*"` for now.